docs/source/testing.rst

```diff
@@ -0,0 +1,316 @@
+
+  def get_all_constraints() -> List[Constraint]:
+      all_requirements = get_basic_requirements() + get_advanced_requirements()
+      return [constraint for requirement in all_requirements for constraint in requirement]
+
+  @pytest.mark.basic
+  def test_basic_constraint(basic_constraint: Constraint, datajudge_engine):
+      test_result = basic_constraint.test(datajudge_engine)
+      assert test_result.outcome, test_result.failure_message
+
+  @pytest.mark.all
+  def test_all_constraint(constraint: Constraint, datajudge_engine):
+      test_result = constraint.test(datajudge_engine)
+      assert test_result.outcome, test_result.failure_message
+
+Once these changes are taken care of, one may run
+
+.. code-block:: console
+
+  $ pytest specification.py -m basic
+
+to only test the basic ``Requirement`` s or
+
+.. code-block:: console
+
+  $ pytest specification.py -m all
+
+to test all ``Requirement`` s.
+
+
+Parametrization
+---------------
+
+A given specification might rely on identifiers such as database names or
+table names. Moreover it might be that, e.g. when iterating from one generation
+of the data to another, these identifiers change. All the while the expectations
+towards how the data relates to another might remain.
+
+In other words, it could be that the logic of the specification should remain
+unchanged while its pointers to data might change. One might simply rely on
+hard-coded references to data. Yet, one might just as well parametrize such
+pointers or identifiers.
+
+For the sake of concreteness, we will assume here that we wish to parametrize
+two identifiers:
+
+* ``new_db``: the name of the 'new database'
+* ``old_db``: the name of the 'old database'
+
+In light of that we will again adapt pytest's ``conftest.py``:
+
+.. code-block:: python
+  :caption: ``conftest.py``
+
+  def pytest_addoption(parser):
+      parser.addoption("--new_db", action="store", help="name of the new database")
+      parser.addoption("--old_db", action="store", help="name of the old database")
+
+
+  def pytest_generate_tests(metafunc):
+      params = {
+          "db_name_new": metafunc.config.option.new_db,
+          "db_ne_old": metafunc.config.option.old_db,
+      }
+      metafunc.parametrize(
+          "constraint",
+          metafunc.module.get_constraints(params),
+          ids=metafunc.module.idfn,
+      )
+
+Now, we can make the creation of our ``Requirement`` s and ``Constraint`` s
+dependent on these parameters:
+
+.. code-block:: python
+  :caption: ``specification.py``
+
+  def get_requirements(params):
+      between_requirement = BetweenRequirement.from_tables(
+          db_name1=params["old_db"],
+	  db_name2=params["new_db"],
+	  ...
+      )
+      ...
+      return requirements
+
+
+  def get_constraints(params):
+      return [
+	  constraint for requirement in get_requirements(params) for constraint in requirement
+      ]
+
+
+  def idfn(constraint):
```

Comment:
Now, in this example you define it, and we see it can be a very simple function. But its purpose is not explained anywhere, one would have to find that in the `pytest_generate_tests` is being used and then go and look in `pytest`s documentation what does it mean.

Maybe a bit more explanation between prose in the documentation and in code comment might help the reader.